diff --git a/lib/dprint/typescript.js b/lib/dprint/typescript.js
index 39358aa1ce6970897193e75f1a10365f9bfba094..f3a3ce6fb697fa2ec00f52681f6a05bb9c9525da 100644
--- a/lib/dprint/typescript.js
+++ b/lib/dprint/typescript.js
@@ -6,10 +6,10 @@ Object.defineProperty(exports, "__esModule", { value: true });
 exports.format = void 0;
 const debug_1 = __importDefault(require("debug"));
 const fs_1 = __importDefault(require("fs"));
-const path_1 = __importDefault(require("path"));
 const debug = (0, debug_1.default)("eslint:plugin-dprint");
 // Load `dprint-plugin-typescript`.
-const TSPluginPath = path_1.default.join(__dirname, "typescript-0.62.0.wasm");
+const { getPath } = require("@dprint/typescript")
+const TSPluginPath = getPath() //path_1.default.join(__dirname, "typescript-0.62.0.wasm");
 const TSPluginModule = new WebAssembly.Module(fs_1.default.readFileSync(TSPluginPath));
 const TSPluginInstance = new WebAssembly.Instance(TSPluginModule, {});
 const TSPlugin = TSPluginInstance.exports;