diff --git a/node_modules/@playwright/test/lib/third_party/tsconfig-loader.js b/node_modules/@playwright/test/lib/third_party/tsconfig-loader.js
index 5e0f3c6..2356d8d 100644
--- a/node_modules/@playwright/test/lib/third_party/tsconfig-loader.js
+++ b/node_modules/@playwright/test/lib/third_party/tsconfig-loader.js
@@ -80,7 +80,7 @@ function loadSyncDefault(cwd, filename, baseUrl) {
 
 function resolveConfigPath(cwd, filename) {
   if (filename) {
-    const absolutePath = fs.lstatSync(filename).isDirectory() ? path.resolve(filename, "./tsconfig.json") : path.resolve(cwd, filename);
+    const absolutePath = fs.lstatSync(filename).isDirectory() ? path.resolve(filename, "./tsconfig.test.json") : path.resolve(cwd, filename);
     return absolutePath;
   }
 
@@ -93,7 +93,7 @@ function resolveConfigPath(cwd, filename) {
 }
 
 function walkForTsConfig(directory, existsSync = fs.existsSync) {
-  const configPath = path.join(directory, "./tsconfig.json");
+  const configPath = path.join(directory, "./tsconfig.test.json");
 
   if (existsSync(configPath)) {
     return configPath;
