diff --git a/lib/third_party/tsconfig-loader.js b/lib/third_party/tsconfig-loader.js
index 1b5780a..14282f4 100644
--- a/lib/third_party/tsconfig-loader.js
+++ b/lib/third_party/tsconfig-loader.js
@@ -73,8 +73,11 @@ function resolveConfigPath(cwd) {
   const configAbsolutePath = walkForTsConfig(cwd);
   return configAbsolutePath ? path.resolve(configAbsolutePath) : undefined;
 }
+
+const configFileName = `./${process.env["TSCONFIG"] ?? "tsconfig.json"}`;
+
 function walkForTsConfig(directory, existsSync = fs.existsSync) {
-  const configPath = path.join(directory, "./tsconfig.json");
+  const configPath = path.join(directory, configFileName);
   if (existsSync(configPath)) {
     return configPath;
   }
