diff --git a/lib/third_party/tsconfig-loader.js b/lib/third_party/tsconfig-loader.js
index 441a9b6..f3ca8c4 100644
--- a/lib/third_party/tsconfig-loader.js
+++ b/lib/third_party/tsconfig-loader.js
@@ -73,8 +73,9 @@ function resolveConfigPath(cwd) {
   const configAbsolutePath = walkForTsConfig(cwd);
   return configAbsolutePath ? path.resolve(configAbsolutePath) : undefined;
 }
+const configFileName = `./${process.env["TSCONFIG"] ?? "tsconfig.json"}`;
 function walkForTsConfig(directory, existsSync = fs.existsSync) {
-  const tsconfigPath = path.join(directory, "./tsconfig.json");
+  const tsconfigPath = path.join(directory, configFileName);
   if (existsSync(tsconfigPath)) {
     return tsconfigPath;
   }