diff --git a/dist/referencer/Referencer.js b/dist/referencer/Referencer.js
index eedc4d9c1d102e463b65be866a862a52bd1fc291..af54cf08f14ff2466895dee03a74c8dcc36c4d33 100644
--- a/dist/referencer/Referencer.js
+++ b/dist/referencer/Referencer.js
@@ -61,6 +61,8 @@ class Referencer extends Visitor_1.Visitor {
             const variables = lib_1.lib[lib];
             // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
             /* istanbul ignore if */ if (!variables) {
+                // TODO: temp workaround for tsplus
+                if (lib === "tsplus") { continue; }
                 throw new Error(`Invalid value for lib provided: ${lib}`);
             }
             for (const [name, variable] of Object.entries(variables)) {
