diff --git a/dist/referencer/Referencer.js b/dist/referencer/Referencer.js
index 7588225..8f454d5 100644
--- a/dist/referencer/Referencer.js
+++ b/dist/referencer/Referencer.js
@@ -60,6 +60,8 @@ class Referencer extends Visitor_1.Visitor {
         for (const lib of __classPrivateFieldGet(this, _Referencer_lib, "f")) {
             const variables = lib_1.lib[lib];
             /* istanbul ignore if */ if (!variables) {
+                // TODO: temp workaround for tsplus
+                continue;
                 throw new Error(`Invalid value for lib provided: ${lib}`);
             }
             for (const [name, variable] of Object.entries(variables)) {
