diff --git a/dist/index.js b/dist/index.js
index 69126edc70f6abdf3ba0306db22059fc812bfa89..6ea6404fedc663e7b74357503e38c02d141c8ee7 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -55,7 +55,7 @@ exports.processors = {
 };
 const codegen = {
     // @ts-expect-error types are wrong?
-    meta: { fixable: true },
+    meta: { fixable: true, schema: false },
     create(context) {
         const validate = () => {
             const sourceCode = context
@@ -115,7 +115,7 @@ const codegen = {
                     return;
                 }
                 const opts = maybeOptions.right || {};
-                const presets = Object.assign(Object.assign({}, presetsModule), (_a = context.options[0]) === null || _a === void 0 ? void 0 : _a.presets);
+                const presets = Object.assign(Object.assign({}, presetsModule), (_a = context.options[0]) === null || _a === void 0 ? void 0 : typeof _a.presets === "string" ? require(_a.presets) : _a.presets);
                 const preset = typeof (opts === null || opts === void 0 ? void 0 : opts.preset) === 'string' && presets[opts.preset];
                 if (typeof preset !== 'function') {
                     context.report({
@@ -129,7 +129,7 @@ const codegen = {
                 const normalise = (val) => val.trim().replace(/\r?\n/g, os.EOL);
                 const result = (0, Either_1.tryCatch)(() => {
                     const meta = { filename: context.getFilename(), existingContent };
-                    return preset({ meta, options: opts });
+                    return preset({ meta, options: opts }, context);
                 }, err => `${err}`);
                 if (result._tag === 'Left') {
                     context.report({ message: result.left, loc: startMarkerLoc });
diff --git a/dist/presets/index.d.ts b/dist/presets/index.d.ts
index a919405f6aeb9f3d3887d69d10350b46b790cf8e..3122de22e92cdf2d95d8bd881d054b35eb223f0c 100644
--- a/dist/presets/index.d.ts
+++ b/dist/presets/index.d.ts
@@ -4,7 +4,7 @@ export declare type Preset<Options> = (params: {
         existingContent: string;
     };
     options: Options;
-}) => string;
+}, context: any) => string;
 export * from './barrel';
 export * from './custom';
 export * from './empty';
