diff --git a/lib/tsc.js b/lib/tsc.js
index 49a42a27bb0974198146da06ddcac8a25683c697..232692963fef5de6fb0ae01a3fc18a796461c4f9 100644
--- a/lib/tsc.js
+++ b/lib/tsc.js
@@ -11425,7 +11425,7 @@ function isInternalDeclaration(node, sourceFile) {
 // src/compiler/utilities.ts
 var resolvingEmptyArray = [];
 var externalHelpersModuleNameText = "tslib";
-var defaultMaximumTruncationLength = 160;
+var defaultMaximumTruncationLength = 3200;
 var noTruncationMaximumTruncationLength = 1e6;
 function getDeclarationOfKind(symbol, kind) {
   const declarations = symbol.declarations;
diff --git a/lib/tsserver.js b/lib/tsserver.js
index 707a6387a5ec9eba8a44bc9e61622dbcd11cffd5..7d8b2638d90f0a9f6d61282af03d7d19b9a631ad 100644
--- a/lib/tsserver.js
+++ b/lib/tsserver.js
@@ -15183,7 +15183,7 @@ function isInternalDeclaration(node, sourceFile) {
 // src/compiler/utilities.ts
 var resolvingEmptyArray = [];
 var externalHelpersModuleNameText = "tslib";
-var defaultMaximumTruncationLength = 160;
+var defaultMaximumTruncationLength = 3200;
 var noTruncationMaximumTruncationLength = 1e6;
 function getDeclarationOfKind(symbol, kind) {
   const declarations = symbol.declarations;
diff --git a/lib/typescript.js b/lib/typescript.js
index a75d7ba4b66e49e217f6f41780df45ee7f3704ed..d8e495952319489f4a48f4c23c6721ceaacec8f4 100644
--- a/lib/typescript.js
+++ b/lib/typescript.js
@@ -18953,7 +18953,7 @@ ${lanes.join("\n")}
       init_ts2();
       resolvingEmptyArray = [];
       externalHelpersModuleNameText = "tslib";
-      defaultMaximumTruncationLength = 160;
+      defaultMaximumTruncationLength = 3200;
       noTruncationMaximumTruncationLength = 1e6;
       stringWriter = createSingleLineStringWriter();
       getScriptTargetFeatures = /* @__PURE__ */ memoize(
