diff --git a/lib/_tsc.js b/lib/_tsc.js
index 3bcdbf7ae59ec00ee5be22ab3219093e691833d7..bfd6734adb8300358522d055dedebb9be8d44df4 100644
--- a/lib/_tsc.js
+++ b/lib/_tsc.js
@@ -12539,7 +12539,7 @@ function isInternalDeclaration(node, sourceFile) {
 // src/compiler/utilities.ts
 var resolvingEmptyArray = [];
 var externalHelpersModuleNameText = "tslib";
-var defaultMaximumTruncationLength = 160;
+var defaultMaximumTruncationLength = 3200;
 var noTruncationMaximumTruncationLength = 1e6;
 function getDeclarationOfKind(symbol, kind) {
   const declarations = symbol.declarations;
diff --git a/lib/typescript.js b/lib/typescript.js
index 33387eadeb8471d4a59907e145c675898c3b9cbf..e10c7590fcabb98f9b15dec7daa8dff48eca9ef0 100644
--- a/lib/typescript.js
+++ b/lib/typescript.js
@@ -16114,7 +16114,7 @@ function isInternalDeclaration(node, sourceFile) {
 // src/compiler/utilities.ts
 var resolvingEmptyArray = [];
 var externalHelpersModuleNameText = "tslib";
-var defaultMaximumTruncationLength = 160;
+var defaultMaximumTruncationLength = 3200;
 var noTruncationMaximumTruncationLength = 1e6;
 function getDeclarationOfKind(symbol, kind) {
   const declarations = symbol.declarations;
