diff --git a/lib/tsc.js b/lib/tsc.js
index 3ea9087202585e3e7ffeb6627106e060abcbf9f0..93633344e7e0df9c0dd90a9cc0cbc671ba721e94 100644
--- a/lib/tsc.js
+++ b/lib/tsc.js
@@ -12462,7 +12462,7 @@ function isInternalDeclaration(node, sourceFile) {
 // src/compiler/utilities.ts
 var resolvingEmptyArray = [];
 var externalHelpersModuleNameText = "tslib";
-var defaultMaximumTruncationLength = 160;
+var defaultMaximumTruncationLength = 3200;
 var noTruncationMaximumTruncationLength = 1e6;
 function getDeclarationOfKind(symbol, kind) {
   const declarations = symbol.declarations;
diff --git a/lib/typescript.js b/lib/typescript.js
index 0c2c66fb524e13da65b1328ee489bc410dffccf4..e398f3890f79589e49f011259d0279f11d092bcb 100644
--- a/lib/typescript.js
+++ b/lib/typescript.js
@@ -16054,7 +16054,7 @@ function isInternalDeclaration(node, sourceFile) {
 // src/compiler/utilities.ts
 var resolvingEmptyArray = [];
 var externalHelpersModuleNameText = "tslib";
-var defaultMaximumTruncationLength = 160;
+var defaultMaximumTruncationLength = 3200;
 var noTruncationMaximumTruncationLength = 1e6;
 function getDeclarationOfKind(symbol, kind) {
   const declarations = symbol.declarations;
